core/trino-main/src/main/java/io/trino/operator/output/RleAwarePositionsAppender.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.output;
+
+import io.trino.spi.block.Block;
+import io.trino.spi.block.RunLengthEncodedBlock;
+import io.trino.type.BlockTypeOperators.BlockPositionEqual;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.Nullable;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class RleAwarePositionsAppender
+        implements PositionsAppender
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(RleAwarePositionsAppender.class).instanceSize();
+    private static final int NO_RLE = -1;
+
+    private final BlockPositionEqual equalOperator;
+    private final PositionsAppender delegate;
+
+    @Nullable
+    private Block rleValue;
+
+    // NO_RLE means flat state, 0 means initial empty state, positive means rle state and the current rle position count.
+    private int rlePositionCount;
+
+    public RleAwarePositionsAppender(BlockPositionEqual equalOperator, PositionsAppender delegate)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+        this.equalOperator = requireNonNull(equalOperator, "equalOperator is null");
+    }
+
+    @Override
+    public void append(IntArrayList positions, Block source)
+    {
+        // RleAwarePositionsAppender should be used with FlatteningPositionsAppender that makes sure
+        // append is called only with flat block
+        checkArgument(!(source instanceof RunLengthEncodedBlock));
+        switchToFlat();
+        delegate.append(positions, source);
+    }
+
+    @Override
+    public void appendRle(RunLengthEncodedBlock source)
+    {
+        if (source.getPositionCount() == 0) {
+            return;
+        }
+
+        if (rlePositionCount == 0) {
+            // initial empty state, switch to rle state
```

Comment:
`switch to rle state`

You are not really switching to RLE state, you are already in initial RLE state. switching to RLE state would make changing `rlePositionCount` from `NO_RLE` to `source.getPositionCount()`